src/controllerclientimpl.cpp

```diff
@@ -2125,4 +2158,292 @@ void ControllerClientImpl::CreateLogEntries(const std::vector<LogEntry>& logEntr
+    }
+
+    // upgrade the connection to websocket
+    if (_tcpStream) {
+        _tcpStream->handshake(host + ":" + std::to_string(port), "/api/v2/graphql");
+    }
+    if (_unixSocketStream) {
+        _unixSocketStream->handshake(host + ":" + std::to_string(port), "/api/v2/graphql");
+    }
+
+    // encode username and password
+    std::string usernamePassword = clientInfo.username + ":" + clientInfo.password;
+    std::string encodedUsernamePassword;
+    encodedUsernamePassword.resize(boost::beast::detail::base64::encoded_size(usernamePassword.size()));
+    boost::beast::detail::base64::encode(&encodedUsernamePassword[0], usernamePassword.data(), usernamePassword.size());
+    
+    // add basic authorization header
+    std::string connectionInitializationMessage = boost::str(boost::format(R"({"type":"connection_init","payload":{"Authorization":"Basic %s"}})") % encodedUsernamePassword);
+    
+    // initialize the websocket
+    this->_SendMessage(connectionInitializationMessage);
+
+    // read connection_ack from server
+    if (_tcpStream) {
+        _tcpStream->read(_subscriptionBuffer);
+    }
+    if (_unixSocketStream) {
+        _unixSocketStream->read(_subscriptionBuffer);
+    }
+
+    std::string message = boost::beast::buffers_to_string(_subscriptionBuffer.data());
+    if (message.find("connection_ack") == std::string::npos) {
+        throw MUJIN_EXCEPTION_FORMAT("Failed to initialize websocket connection, Expected 'connection_ack' in response, but got: %s", message, MEC_HTTPServer);
+    }
+    _subscriptionBuffer.clear();
+
+    // start the asynchronous read
+    if (_tcpStream) {
+        _ReadFromSubscriptionStream(_tcpStream, &_subscriptionBuffer, _onReadHandlers, &_mutex, _allocator);
+    }
+    if (_unixSocketStream) {
+        _ReadFromSubscriptionStream(_unixSocketStream, &_subscriptionBuffer, _onReadHandlers, &_mutex, _allocator);
+    }
+
+    // start a new thread running I/O service until the socket is closed
+    _thread = boost::make_shared<std::thread>([ioContext] {
+        ioContext->run();
+    });
+}
+
+bool GraphSubscriptionWebSocketHandler::IsStreamOpen()
+{   
+    boost::mutex::scoped_lock lock(_mutex);
+    if (_tcpStream) {
+        return _tcpStream->is_open();
+    }
+    if (_unixSocketStream) {
+        return _unixSocketStream->is_open();
+    }
+}
+
+std::string GraphSubscriptionWebSocketHandler::StartSubscription(const std::string& operationName, const std::string& query, const rapidjson::Value& rVariables, std::function<void(const boost::system::error_code&, rapidjson::Value&&)> onReadHandler)
+{
+    boost::mutex::scoped_lock lock(_mutex);
+
+    // generate a random id for the subsctiption
+    std::string subscriptionId = boost::uuids::to_string(_randomGenerator());
+    MUJIN_LOG_INFO(boost::format("subscription %s started") % subscriptionId);
+
+    // build the query
+    rapidjson::Value payload;
```

Comment:
rapidjson objects start with `r`, please follow the naming conventions in this file

```suggestion
    rapidjson::Value rPayload;
```